OSDOCS-20096 added assembly/modules for spiffe helper (#112997)
openshift/openshift-docs: modules/zero-trust-manager-deploy-spiffe-helper.adoc

The stock SPIFFE Helper image does not include the `psql` client required for that reload command. Build a custom image that extends the {zero-trust-full} image and adds the PostgreSQL client package.

.Prerequisites

* You have access to the cluster with `cluster-admin` privileges.
* You have installed the {zero-trust-full}.
* The `ZeroTrustWorkloadIdentityManager` custom resource reports `Ready`.

.Procedure

. Create a `Containerfile` with the following contents:
+
[source,dockerfile]
----
FROM registry.redhat.io/zero-trust-workload-identity-manager/spiffe-helper-rhel9:<version>

USER 0

RUN dnf install -y postgresql && \
    dnf clean all

USER 65534
----
+
Replace `<version>` with the tag that matches your {zero-trust-full} installation.

. Build and push a custom SPIFFE Helper image for the PostgreSQL server by running the following command.
+
[source,terminal]
----
$ podman build -f Containerfile \
  -t <registry>/spiffe-helper-postgresql:latest .
$ podman push <registry>/spiffe-helper-postgresql:latest
----
+
Replace `<registry>` with your container registry.

> . Deploy the PostgreSQL server resources by running the following command:

Reviewer: ```suggestion
. Create the PostgreSQL server resources by using the following example:
```

Author: Done

+
[source,yaml]
----

> $ oc apply -f - << 'EOF'

Reviewer: ```suggestion
```
having the command embedded in the yamls is breaking the formatting

Author: I put it through a YAML validator and it passed

Reviewer: this breaks the formatting and highlighting on our **website**

Author: We already have formatting like this in our document [Creating the demonstration application](https://github.com/openshift/openshift-docs/pull/112274)

Author: [Deploying the workload application](https://docs.redhat.com/en/documentation/openshift_container_platform/4.21/html-single/security_and_compliance/index#zero-trust-manager-deploy-app_zero-trust-manager-oidc-federation)

Reviewer: at some point, come back and separate the command from the yaml if you can; it messes up the website formatting and support prefers everything separate

    matchLabels:
      kubernetes.io/metadata.name: postgresql-spiffe
  podSelector:
    matchLabels:
      app: postgresql-spiffe
  spiffeIDTemplate: 'spiffe://{{ .TrustDomain }}/ns/{{ .PodMeta.Namespace }}/sa/{{ .PodSpec.ServiceAccountName }}'
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: postgres-config
  namespace: postgresql-spiffe
data:
  pg_hba.conf: |
    # TYPE      DATABASE        USER            ADDRESS                 METHOD
    local       all             all                                     trust
    host        postgres        postgres        127.0.0.1/32            trust
    hostnossl   postgres        postgres        127.0.0.1/32            trust
    hostnossl   all             all             0.0.0.0/0               reject
    hostssl     all             all             0.0.0.0/0               cert
  postgresql.conf: |
    listen_addresses '*'
    ssl = on
    ssl_cert_file = '/opt/postgresql-certs/svid.pem'
    ssl_key_file = '/opt/postgresql-certs/svid.key'
    ssl_ca_file = '/opt/postgresql-certs/svid_bundle.pem'
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: postgresql-init-db
  namespace: postgresql-spiffe
data:
  initdb.sh: |

>     # Copy configuration files

Reviewer: ```suggestion
```

Author: Done

Author: these can stay

    cp /opt/pg_hba/pg_hba.conf /var/lib/pgsql/data/userdata

>     # Create postgresql resources

Reviewer: ```suggestion
```
i can't be sure, but wondering if these commented-out footnotes are causing the formatting issues below

Author: When I put it into a yaml validator, it says it's valid yaml

Author: This should now be valid YAML. Formats like this have been used in the past in zero-trust and seemed to work

Reviewer: this is not docs style, though; i'll dig around and see if i can find the reference

            - '-config'
            - /etc/spiffe-helper/helper.conf
          volumeMounts:
            - name: spiffe-workload-api
              readOnly: true
              mountPath: /spiffe-workload-api
            - name: postgresql-certs
              mountPath: /opt/postgresql-certs
            - name: spiffe-helper
              mountPath: /etc/spiffe-helper
          securityContext:
            capabilities:
              drop:
                - ALL
      volumes:
        - name: spiffe-workload-api
          csi:
            driver: csi.spiffe.io
            readOnly: true
        - name: postgresql-certs
          emptyDir:
            medium: Memory
        - name: spiffe-helper
          configMap:
            name: spiffe-helper
        - name: postgresql-init-db
          configMap:
            name: postgresql-init-db
        - name: pg-hba
          configMap:
            name: postgres-config
            items:
              - key: pg_hba.conf
                path: pg_hba.conf
        - name: postgres-config
          configMap:
            name: postgres-config
            items:
              - key: postgresql.conf
                path: postgresql.conf

> EOF

Reviewer: ```suggestion
```

Author: Made changes will need to send back for review

Author: The EOF can stay

Reviewer 2: EOF must stay

----
+
Replace both `<registry>` occurrences in the deployment with your container registry.

> 

Reviewer: ```suggestion
. Apply the PostgreSQL server resources by running the following command:
+
[source,terminal]
----
$ oc apply -f psql-server-resources.yaml
----
```

just an example; not sure what a user would call this config

Author: Since I'm keeping the YAML as is, this is not needed

. Confirm that the SPIFFE Helper configuration matches the PostgreSQL TLS settings by running the following command:
+
[source,terminal]
----
$ oc get configmap spiffe-helper -n postgresql-spiffe \
  -o jsonpath='{.data.helper\.conf}{"\n"}'
----
+
The `cert_dir` and `svid_*` file names must match the `ssl_*_file` paths in `postgresql.conf`. The server `helper.conf` must include settings similar to the following example:
+
[source,terminal]
----
agent_address = "/spiffe-workload-api/spire-agent.sock"
cmd = "/usr/bin/psql"
cmd_args = "-U postgres -h 127.0.0.1 -c \"SELECT pg_reload_conf();\""
cert_dir = "/opt/postgresql-certs"
renew_signal = ""
svid_file_name = "svid.pem"
svid_key_file_name = "svid.key"
svid_bundle_file_name = "svid_bundle.pem"
----

. Deploy the PostgreSQL client by running the following command:
+
The client uses the stock SPIFFE Helper image from the {zero-trust-full}. Only the server deployment requires the custom image with the `psql` client.
+
[source,yaml]
----

> $ oc apply -f - << 'EOF'

Reviewer: same thing here, this should be a config and and an apply (two steps)

Author: Actually that's valid YAML. It's been used before

Reviewer: exactly; the website translation sees a yaml block with a terminal command, so the box type for web rendering is mixed

            - name: postgresql-certs
              mountPath: /opt/postgresql-certs
          securityContext:
            capabilities:
              drop:
                - ALL
        - name: spiffe-helper
          image: registry.redhat.io/zero-trust-workload-identity-manager/spiffe-helper-rhel9:<version>
          args:
            - '-config'
            - /etc/spiffe-helper/helper.conf
          volumeMounts:
            - name: spiffe-workload-api
              readOnly: true
              mountPath: /spiffe-workload-api
            - name: postgresql-certs
              mountPath: /opt/postgresql-certs
            - name: spiffe-helper
              mountPath: /etc/spiffe-helper
          securityContext:
            capabilities:
              drop:
                - ALL
      volumes:
        - name: spiffe-workload-api
          csi:
            driver: csi.spiffe.io
            readOnly: true
        - name: postgresql-certs
          emptyDir:
            medium: Memory
        - name: spiffe-helper
          configMap:
            name: spiffe-helper
EOF
----
+
Replace `<version>` with the tag that matches your {zero-trust-full} installation.

> . Confirm that pods are running in both namespaces by running the following commands:

Reviewer: should be two steps

+
[source,terminal]
----
$ oc get pods -n postgresql-spiffe
$ oc get pods -n postgresql-spiffe-client
----

.Verification

. Inspect the server certificate by running the following command:
+
[source,terminal]
----
$ oc rsh -n postgresql-spiffe -c postgresql-spiffe \
  deployment/postgresql-spiffe \
  cat /opt/postgresql-certs/svid.pem | openssl x509 -noout -text
----
+

> The certificate includes the service hostname `postgresql-spiffe.postgresql-spiffe.svc`.

Reviewer: is this what i expect in the output? i don't understand this sentence

Author: Done

. Inspect the client certificate by running the following command:
+
[source,terminal]
----
$ oc rsh -n postgresql-spiffe-client -c postgresql-spiffe-client \
  deployment/postgresql-spiffe-client \
  cat /opt/postgresql-certs/svid.pem | openssl x509 -noout -text
----
+

> The certificate CN should be `postgresql_spiffe`.

Reviewer: what's CN? define or backtick, perhaps?

Author: Done

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-spiffe-helper.adoc

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-deploy-spiffe-helper_{context}"]
= Deploying PostgreSQL with SPIFFE Helper

[role="_abstract"]
Deploy a PostgreSQL server and client that use SPIFFE Helper to fetch X.509 {svid-full} from the SPIFFE Workload API and write them to disk for mutual Transport Layer Security (mTLS).

The manifests in this procedure use the {zero-trust-full} SPIFFE Helper image.

> When certificates renew, the server must reload PostgreSQL TLS configuration. Containers in the same pod use separate process namespaces, so the SPIFFE Helper sidecar cannot signal the PostgreSQL container directly. Instead, SPIFFE Helper uses the *managed-child* pattern: in daemon mode it runs `cmd` (`/usr/bin/psql`) with `cmd_args` to execute `SELECT pg_reload_conf();` as a child process inside the helper container whenever X.509 material is updated.

Reviewer 3: ```suggestion
When certificates renew, the server must reload PostgreSQL TLS configuration. Containers in the same pod use separate process namespaces, so the SPIFFE Helper sidecar cannot signal the PostgreSQL container directly. Instead, SPIFFE Helper uses the *managed-child* pattern. This pattern in daemon mode runs `cmd` (`/usr/bin/psql`) with `cmd_args` to execute `SELECT pg_reload_conf();` as a child process inside the helper container whenever X.509 material is updated.
```

Author: Done

The stock SPIFFE Helper image does not include the `psql` client required for that reload command. Build a custom image that extends the {zero-trust-full} image and adds the PostgreSQL client package.

.Prerequisites

* You have access to the cluster with `cluster-admin` privileges.
* You have installed the {zero-trust-full}.
* The `ZeroTrustWorkloadIdentityManager` custom resource reports `Ready`.

.Procedure

. Create a `Containerfile` with the following contents:
+
[source,dockerfile]
----
FROM registry.redhat.io/zero-trust-workload-identity-manager/spiffe-helper-rhel9:<version>

USER 0

RUN dnf install -y postgresql && \
    dnf clean all

USER 65534
----

> +

Reviewer 3: <img width="1138" height="325" alt="Image" src="https://github.com/user-attachments/assets/f22ac38b-ff31-47d3-8260-984ce69ec6fd" />

Author: Done
